gtk/src/lib.rs

```diff
@@ -26,55 +31,75 @@ use std::{
     thread,
 };
 
+// const RECOVERY_PARTITION: usize = 0;
+const REFRESH_OS: usize = 0;
+
 pub type ErrorCallback = Rc<RefCell<Box<dyn Fn(&str)>>>;
 pub type EventCallback = Rc<RefCell<Box<dyn Fn(Event)>>>;
 pub type ReadyCallback = Rc<RefCell<Box<dyn Fn()>>>;
 
 #[derive(Shrinkwrap)]
 pub struct UpgradeWidget {
-    sender: mpsc::SyncSender<BackgroundEvent>,
+    sender:         mpsc::SyncSender<BackgroundEvent>,
     callback_error: ErrorCallback,
     callback_event: EventCallback,
     callback_ready: ReadyCallback,
     #[shrinkwrap(main_field)]
-    container: gtk::Container,
+    container:      gtk::Container,
 }
 
-impl UpgradeWidget {
-    pub fn new() -> Self {
+impl Default for UpgradeWidget {
+    fn default() -> Self {
         let (bg_sender, bg_receiver) = mpsc::sync_channel(5);
         let (gui_sender, gui_receiver) = glib::MainContext::channel(glib::PRIORITY_DEFAULT);
         let gui_sender = Arc::new(gui_sender);
 
-        {
-            let gui_sender = gui_sender.clone();
-
-            thread::spawn(move || {
-                events::background::run(bg_receiver, move |event| {
-                    let _ = gui_sender.send(event);
-                });
+        thread::spawn(enclose!((gui_sender) move || {
+            events::background::run(bg_receiver, move |event| {
+                let _ = gui_sender.send(event);
+            });
+        }));
+
+        let button_sg = gtk::SizeGroup::new(gtk::SizeGroupMode::Both);
+        let option_sg = gtk::SizeGroup::new(gtk::SizeGroupMode::Both);
+        let sublab_sg = gtk::SizeGroup::new(gtk::SizeGroupMode::Both);
+
+        let mut upgrade = Section::new("<b>OS Upgrade</b>");
+        let mut recovery = Section::new("<b>OS Recovery</b>");
+
+        upgrade.add_option(&option_sg, &button_sg, &sublab_sg, |option| {
+            option.button_class(&gtk::STYLE_CLASS_SUGGESTED_ACTION);
+        });
+
+        recovery
+            // .add_option(&option_sg, &button_sg, &sublab_sg, |option| {
+            //     option
+            //         .button_class(&gtk::STYLE_CLASS_SUGGESTED_ACTION)
+            //         .label("Recovery Partition")
+            //         .sublabel(Some("You have the most current version of the recovery version"));
+            // })
+            .add_option(&option_sg, &button_sg, &sublab_sg, |option| {
+                option
+                    .button_label("Refresh")
+                    .label("Refresh OS")
+                    .sublabel(Some("Reinstall while keeping user accounts and files"));
             });
-        }
-
-        let upgrade = Section::new("<b>OS Upgrade</b>");
 
         let dismisser = gtk::ListBoxRow::new();
         upgrade.list.add(&dismisser);
 
         let button_sg = cascade! {
             gtk::SizeGroup::new(gtk::SizeGroupMode::Both);
-            ..add_widget(&upgrade.option.button);
+            ..add_widget(&upgrade.options[0].button);
+            ..add_widget(&recovery.options[0].button);
```

Comment:
Should this be using `REFRESH_OS` for the index?

Reply:
Yes, it should be explicitly declared here.